pkg/distributor/distributor.go

```diff
@@ -206,17 +215,23 @@ func (d *Distributor) Push(ctx context.Context, req *logproto.PushRequest) (*log
 	validatedSamplesCount := 0
 
 	for _, stream := range req.Streams {
-		ls, err := logql.ParseLabels(stream.Labels)
-		if err != nil {
-			validationErr = httpgrpc.Errorf(http.StatusBadRequest, "error parsing labels: %v", err)
-			continue
-		}
-		// ensure labels are correctly sorted.
-		// todo(ctovena) we should lru cache this
-		stream.Labels = ls.String()
-		if err := d.validator.ValidateLabels(userID, ls, stream); err != nil {
-			validationErr = err
-			continue
+		label, ok := d.labelCache.Get(stream.Labels)
+		if !ok {
+			ls, err := logql.ParseLabels(stream.Labels)
+			if err != nil {
+				validationErr = httpgrpc.Errorf(http.StatusBadRequest, "error parsing labels: %v", err)
+				continue
+			}
+			// ensure labels are correctly sorted.
+			// todo(ctovena) we should lru cache this
+			stream.Labels = ls.String()
+			if err := d.validator.ValidateLabels(userID, ls, stream); err != nil {
+				validationErr = err
+				continue
+			}
+			d.labelCache.Add(stream.Labels, stream.Labels)
+		} else {
+			stream.Labels = label.(string)
```

Comment:
I think it would be good to move this to a separate function and write a benchmark for it.